application/frontend/controllers/MethodController.php

```diff
@@ -38,6 +40,159 @@ public function actionIndex()
+        if ($method !== null) {
+            return $method;
+        }
+
+        /*
+         * Create method entry to be verified
+         */
+        $newMethod = Method::createAndSendVerification(
+            \Yii::$app->user->getId(),
+            $type,
+            $value
+        );
+
+        return $newMethod;
+    }
+
+    /**
+     * Validates user submitted code and marks method as verified if valid
+     * @param string $uid
+     * @return Method
+     * @throws BadRequestHttpException
+     * @throws NotFoundHttpException
+     * @throws TooManyRequestsHttpException
+     * @throws \Exception
+     */
+    public function actionUpdate($uid)
+    {
+        /*
+         * Delete methods not yet verified that have expired - Just-in-time cleanup
+         */
+        try {
+            Method::deleteExpiredUnverifiedMethods();
+        } catch (\Exception $e) {
+            \Yii::error([
+                'action' => 'deleteExpiredUnverifiedMethods',
+                'status' => 'error',
+                'error' => $e->getMessage(),
+            ]);
+        }
+
+        /*
+         * Find method if belongs to user and is not expired
+         */
+        /** @var Method $method */
+        $method = Method::findOne([
+            'uid' => $uid,
+            'user_id' => \Yii::$app->user->getId(),
+        ]);
+
+        /*
+         * If not found, throw 404
+         */
+        if ($method === null) {
+            throw new NotFoundHttpException();
+        }
+
+        /*
+         * If method is already verified, just return it as if successful
+         */
+        if ($method->verified === 1) {
+            return $method;
+        }
+
+        /*
+         * Ensure verification attempts is not above limit
+         */
+        if ($method->verification_attempts >= \Yii::$app->params['reset']['maxAttempts']) {
+            throw new TooManyRequestsHttpException();
+        }
+
+        /*
+         * Get verification code and attempt to verify
+         */
+        $request = \Yii::$app->request;
+        $code = $request->getBodyParam('code');
+        if ($code === null) {
+            throw new BadRequestHttpException('Code is required');
+        }
+
+        try {
+            $method->validateAndSetAsVerified($code);
+        } catch (\Exception $e) {
+            throw new BadRequestHttpException('Invalid verification code');
+        }
+
+        return $method;
+    }
+
+    /**
+     * Delete method
+     * @param string $uid
+     * @return array
+     * @throws NotFoundHttpException
+     * @throws \Exception
+     */
+    public function actionDelete($uid)
+    {
+        /** @var Method $method */
+        $method = Method::findOne([
+            'uid' => $uid,
+            'user_id' => \Yii::$app->user->getId()
+        ]);
+
+        if ($method === null) {
+            throw new NotFoundHttpException();
+        }
+
+        $method->delete();
```

Comment:
Yes, I wish the model methods just threw exceptions themselves. I added a check and exception for failure.